File: src/main/GO/leetcode/lists/hot100/LeetCode309.go
package hot100

func maxProfit309(prices []int) int {
	if len(prices) == 0 {
		return 0
	}
	dp := make([][]int, len(prices))
	for i := 0; i < len(dp); i++ {
		dp[i] = make([]int, 2)
	}
	dp[0][0] = 0          //未持有
	dp[0][1] = -prices[0] //持有
	dp[0][2] = 0          //未持有,但是在冷冻期
	for i := 1; i < len(prices); i++ {
		dp[i][0] = Max(dp[i-1][0], dp[i-1][2])
		dp[i][1] = Max(dp[i-1][0]-prices[i], dp[i-1][1])
		dp[i][2] = dp[i-1][1] + prices[i]
	}
	return Max(dp[len(prices)-1][0], dp[len(prices)-1][2])
}

File: src/main/java/leetcode/lists/hot100/LeetCode309.java
package leetcode.lists.hot100;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/5 22:32
 */
public class LeetCode309 {

    public int maxProfit(int[] prices) {

        if (prices == null || prices.length == 0) {
            return 0;
        }

        int[][] dp = new int[prices.length][3];
        dp[0][0] = -prices[0];
        // 0 手上有股票
        // 1 手上没有股票
        // 2 手上没有股票，且在冷冻期
        for (int i = 1; i < prices.length; i++) {
            dp[i][0] = Math.max(dp[i - 1][1] - prices[i], dp[i - 1][0]);
            //fix
            dp[i][1] = Math.max(dp[i - 1][1], dp[i - 1][2]);
            //fix
            dp[i][2] = dp[i - 1][0] + prices[i];
        }
        //fix
        return Math.max(dp[dp.length - 1][1], dp[dp.length - 1][2]);
    }
}
